### PR TITLE
Continuations: Use Write-once memory from STD

### PR DESCRIPTION
**Pull request by georgwiese**
Depends on #1272

Uses the write-once memory introduced in #1202 in our RISC-V machine with continuations.

### CODE REVIEW

File: riscv/src/continuations/bootloader.rs [hunk 1]
```diff
@@ -78,22 +78,22 @@ pub fn bootloader_preamble() -> String {
     for (i, reg) in REGISTER_NAMES.iter().enumerate() {
         let reg = reg.strip_prefix("main.").unwrap();
         preamble.push_str(&format!(
-            "    public initial_{reg} = bootloader_input_value({i});\n"
+            "    public initial_{reg} = main_bootloader_inputs.value({i});\n"
```
Comment (georgwiese) [inline riscv/src/continuations/bootloader.rs:81]: This is a bit unfortunate still! Here the main machine exposes cells in the memory machine as public outputs. Not sure what's a great way to clean this up!

File: riscv/src/continuations/bootloader.rs [hunk 2]
```diff
@@ -50,18 +50,18 @@ pub fn bootloader_preamble() -> String {
...
+    std::write_once_memory::WriteOnceMemory bootloader_inputs;
+
+    instr load_bootloader_input X -> Y = bootloader_inputs.access X, Y ->;
+    instr assert_bootloader_input X, Y -> = bootloader_inputs.access X, Y ->;
 
     let tmp_bootloader_value;
 
     // Sets the PC to the bootloader input at the provided index if it is nonzero
+    // TODO: This should work, but this leads to a wrong PC update rule.
+    // instr jump_to_bootloader_input X = bootloader_inputs.access X, pc' ->;
```
Comment (georgwiese) [inline riscv/src/continuations/bootloader.rs:62]: #1204

File: riscv/src/continuations/bootloader.rs (hunk 1 above)
Comment (georgwiese) [reply riscv/src/continuations/bootloader.rs:81]: OK, this also seems to cause problems with PIL Verifier somehow, it doesn't find the publics! (See failing test) Creating an eStark proof works though!

File: riscv/src/continuations/bootloader.rs [hunk 3]
```diff
@@ -52,20 +52,13 @@ pub fn bootloader_preamble() -> String {
...
     // Write-once memory
-    let BOOTLOADER_INPUT_ADDRESS: col = |i| i;
-    let bootloader_input_value;
-    // Loads a value. If the cell is empty, the prover can choose a value.
-    instr load_bootloader_input X -> Y { {X, Y} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value} }
-    instr assert_bootloader_input X, Y { {X, Y} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value} }
-
-    let tmp_bootloader_value;
-
-    // Sets the PC to the bootloader input at the provided index if it is nonzero
```
Comment (georgwiese) [inline riscv/src/continuations/bootloader.rs:63]: Removed "if it is nonzero". As far as I can see, this was not the case before and it is not now.

**Comment by leonardoalt** [comment]
needs a merge from main?

File: riscv/src/continuations/bootloader.rs [hunk 4]
```diff
@@ -52,20 +52,13 @@ pub fn bootloader_preamble() -> String {
...
-        {X, tmp_bootloader_value} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value},
-        pc' = tmp_bootloader_value
-    }
+    std::write_once_memory::WriteOnceMemory bootloader_inputs;
+
+    instr load_bootloader_input X -> Y = bootloader_inputs.access X, Y ->;
+    instr assert_bootloader_input X, Y -> = bootloader_inputs.access X, Y ->;
+
+    // Sets the PC to the bootloader input at the provided index
+    instr jump_to_bootloader_input X = bootloader_inputs.access X, pc' ->;
```
Comment (leonardoalt) [inline riscv/src/continuations/bootloader.rs:61]: wait, this works with `pc'`? damn

**Review by leonardoalt — APPROVED** [review]
looks good, any performance differences?

**Comment by georgwiese** [comment]
Didn't measure, but should compile to the same thing! Except that it saves one witness column because we write to `pc'` directly now.

**Pull request merged**